---
title: "consensus_tax_mostCom Demo"
author: "K Son"
date: "April 30, 2020"
output: github_document
---

## Overview

Here, we step through implementations of the consensus_tax_mostCom.R algorithm to demonstrate proper uses and anticipated results. This algorithm is built to utilize information from multiple taxonomy tables (defined here as a table of ASVs and correspnding taxonomic assignments) in order to improve the resolution and/or accuracy of taxonomic annotations of amplicon sequences. It incorporates information from multiple taxonomy tables and determines a "consensus taxonomy" for each ASV in your data set. The algorithm requires that all taxonomy tables follow the same taxonomic naming and ranking conventions, that the order of columns in each taxonomy table follows the taxonomic ranking heirarchy, and that the order of rows (ASVs) in each of the input taxonomy tables is the same. If these rules are not followed, spurious results are likely.

### consensus_tax_mostCom Algorithm Description

consensus_tax_mostCom generates a consensus taxonomy by using the most frequent taxonomy starting at the most specific ranking across all user-specified input taxonomy tables. In other words, it's designed to get the the taxonomic annotation where it is the majority among other taxonomic assignments. Unresolved taxonomic assignments in each taxonomy table should be indicated by NA as there is a parameter for the algorithm to consider NA as a taxonomic assignment or not. There will be occasions when multiple input taxonomy tables are 'tied' for the most frequeny taxonomy for a given ASV; thus, after assigning a consensus taxonomy to the ASVs where a single taxonomy provides the most common taxonomy, the algorithm uses a series of user-speicifed rules to break the remaining ties. Rules you may specify are demonstarted below. 

### Setting Things Up:

We'll clear out our environment, load in the reshape2 package ofr later, set our wd, and read in taxonomy tables: The taxonomy tables used here come from implementations of the RDP Bayesian classifer, the new idtaxa algorithm, and MEGAN'S LCA algorithm against the pr2 reference databases. Our amplicon data set is an 18S-V9 tag sequencing project from the coastal ocean. 

You can do this with any taxonomy tables assuming you format them properly. To follow along with this demo, grab the taxonomy tables in the "test_data" directory of this repository and follow the code below.

```{r, warning=FALSE, message=FALSE}
library(dplyr)

rm(list = ls())

# setwd and read in your datasets:
setwd("~/Desktop/Taxonomic Sequencing/amplicon_bioinformatics/tax_pipe_Mar20/")
dd <- "~/Desktop/Taxonomic Sequencing/amplicon_bioinformatics/package_deal/"

# tax mapping:
source(paste0(dd,"taxmapper.R"))

# helper fcns:
source(paste0(dd,"bayestax2df.R"))
source(paste0(dd,"idtax2df_pr2.R"))

bayes.pr2 <- readRDS("initial_tax_tabs/bayes_pr2_0boot_Mar20.rds")
idtax.pr2 <- readRDS("initial_tax_tabs/idtax_pr2_0boot_Mar20.rds")
lca.pr2 <- readRDS("initial_tax_tabs/LCA_pr2_rawdf_Mar20.rds")
```

### Arranging and formating our taxonomy tables for running the algorithm:

```{r, warning=FALSE, message=FALSE}
# here's the rubric for aligning ASV numbers and sequences across datasets:
library("DECIPHER")

setwd("~/Desktop/Taxonomic Sequencing/amplicon_bioinformatics/tax_pipe_Mar20/")
rubber <- readDNAStringSet("blaster/allASVs4blast.fasta")

# convert idtax and bayes to dataframes - keep boot = 0  and return conf for now...
# note that for idtaxa, order of ASVs must be retrieved from dada2 seqtab - rubric represents that order
# and adds ASV names and sequences to output dataframe:
xx <- bayestax2df(bayes.pr2, boot = 0, rubric = rubber, return.conf = TRUE)
bayes.pr2 <- xx[[1]]
bayes.pr2.conf <- xx[[2]]

xx <- idtax2df_pr2(idtax.pr2, boot = 0, rubric = rubber, return.conf = TRUE)
idtax.pr2 <- xx[[1]]
idtax.pr2.conf <- xx[[2]]

# re-order your 6 tables by sorting according to ASV:
ii <- base::sort(bayes.pr2$ASV, index.return = TRUE)
bayes.pr2 <- bayes.pr2[ii$ix,]
bayes.pr2.conf <- bayes.pr2.conf[ii$ix,]
ii <- base::sort(idtax.pr2$ASV, index.return = TRUE)
idtax.pr2 <- idtax.pr2[ii$ix,]
idtax.pr2.conf <- idtax.pr2.conf[ii$ix,]
ii <- base::sort(lca.pr2$ASV, index.return = TRUE)
lca.pr2 <- lca.pr2[ii$ix,]
```

You can run this for a sanity check:

```{r, eval=FALSE}
# check that they're all in the same order
identical(bayes.pr2$ASV, idtax.pr2$ASV)
```

...and this to see what the data sets look like. These data sets are available in the test-data directory.

```{r, eval=FALSE}
head(bayes.pr2)
head(idtax.pr2)
head(lca.pr2)
```

The algorithm expects that the ASV's in each taxonomy data frame are all in the same order. Before inputting tables in the algorithm, we recommend standardizing the taxonomy tables to one refernece database using the taxonomy mapping function as well as sort them by svN and ASV's.


We've created a test data set where we will read in that showcases the expected output of our algorithm. Within the test data set, it stores the taxonomic assignments from each taxonomic algoirthm for each ASV and its corresponding expected consensus result counting non-assignments as a majority candidate. In order to show different scenarios of our algorithm resolving the majorities, we've handpicked certain rows to build a subset data frame to showcase our algorithm. 

```{r}
library(readxl)
consensus.res <- read_excel("~/Desktop/Taxonomic Sequencing/amplicon_bioinformatics/taxonomy_pipeline/test_data/bayes_idtax_lca_pr2_mapped.xlsx")

make.true.NA <- function(x) if(is.character(x)||is.factor(x)){
    is.na(x) <- x=="NA"; x} else {
      x}

consensus.res[] <- lapply(consensus.res, make.true.NA)

colnames(consensus.res)[12] <- "table"

head(consensus.res)
```


### First run

Our data should be good to go, so let's run the algorithm. We have to specify names for our taxonmy tables in a character vector. There is an option for the algorithm to count the term NA as part of the majority computation. In this case, we'll count it. 

First we'll tell R where to find the algorithm and load it into our session. In our first run, we'll specify no tie-breakers, so that the algorithm will only assign consensus taxonomies to ASVs where a single taxonomy table has the most common taxonomy of our 3 datasets.

The output is just a consensus taxonomy table.

Here, we'll pop out a few good examples. 

```{r, warning=FALSE}
source("~/Desktop/Taxonomic Sequencing/amplicon_bioinformatics/taxonomy_pipeline/consensus_taxonomies/consensus_tax_mostCom.R")

cbind(rbind(as.data.frame(consensus.res[1, c(1:2, 4:11)]),
      cbind(consensus.res[1, 1:2], consensus.res[2, 4:11]),
      cbind(consensus.res[1, 1:2], consensus.res[3, 4:11]),
      consensus_tax_mostCom(as.data.frame(consensus.res[1, c(1:2, 4:11)]), 
                       cbind(consensus.res[1, 1:2], consensus.res[2, 4:11]), 
                       cbind(consensus.res[1, 1:2], consensus.res[3, 4:11]),
                       tablenames=c('bayes', 'idtax', 'lca'), 
                       tiebreakz='none', count.na=TRUE)),consensus.res[1:4, 12])
```

The output shows the taxonomic assignments for an arbitrary ASV in 3 of our initial taxonomy tables, as well as our consensus table. The last column shows the taxonomy table of each row for this arbitrary ASV. RESULT deontes our consensus taxonomy. We can see that all of the taxonomies are the same, so it's obvious that the resulting consensus taxonomy should be the same as well.

Now let's look at one where some contain different taxonomies.

```{r, warning=FALSE}
cbind(rbind(as.data.frame(consensus.res[11, c(1:2, 4:11)]),
      cbind(consensus.res[11, 1:2], consensus.res[12, 4:11]),
      cbind(consensus.res[11, 1:2], consensus.res[13, 4:11]),
      consensus_tax_mostCom(as.data.frame(consensus.res[11, c(1:2, 4:11)]), 
                       cbind(consensus.res[11, 1:2], consensus.res[12, 4:11]), 
                       cbind(consensus.res[11, 1:2], consensus.res[13, 4:11]),
                       tablenames=c('bayes', 'idtax', 'lca'), 
                       tiebreakz='none', count.na=TRUE)),consensus.res[11:14, 12])
```

Here, we can see that the taxonomies in bayes and idtax are the majority in each column compared to the NA taxonomy in LCA. Therefore, the resulting consensus taxonomy would be the taxonomies from bayes and idtax. 

```{r, warning=FALSE}
cbind(rbind(as.data.frame(consensus.res[81, c(1:2, 4:11)]),
      cbind(consensus.res[81, 1:2], consensus.res[82, 4:11]),
      cbind(consensus.res[81, 1:2], consensus.res[83, 4:11]),
      consensus_tax_mostCom(as.data.frame(consensus.res[81, c(1:2, 4:11)]), 
                       cbind(consensus.res[81, 1:2], consensus.res[82, 4:11]), 
                       cbind(consensus.res[81, 1:2], consensus.res[83, 4:11]),
                       tablenames=c('bayes', 'idtax', 'lca'), 
                       tiebreakz='none', count.na=TRUE)),consensus.res[81:84, 12])
```

We can see that in the Order, Family, and Genus rankings, there is no majority as there is a three way tie between the taxonomy tables. Since we didn't specify a tiebreaker preference, the algorithm automatically assigns NA to that column as it wasn't able to conclude to a consensus. 

### Second run

Here we will introduce how our algorithm takes care of tie breakers set by the user input. Let's say we want to prioritize the bayes taxonomy table as the tie breaker. This is specified by introducing a list of pairs of (table name, taxonomy) in the order of priority. This means that the first element in the list will be prioritized the most while the last element in thel ist will be prioritized the least.

```{r, warning=FALSE}
cbind(rbind(as.data.frame(consensus.res[81, c(1:2, 4:11)]),
      cbind(consensus.res[81, 1:2], consensus.res[82, 4:11]),
      cbind(consensus.res[81, 1:2], consensus.res[83, 4:11]),
      consensus_tax_mostCom(as.data.frame(consensus.res[81, c(1:2, 4:11)]), 
                       cbind(consensus.res[81, 1:2], consensus.res[82, 4:11]), 
                       cbind(consensus.res[81, 1:2], consensus.res[83, 4:11]),
                       tablenames=c('bayes', 'idtax', 'lca'), 
                       tiebreakz=list(c('bayes', NA)), count.na=TRUE)),consensus.res[81:84, 12])
```

We can see that whenever there is a three way tie, the taxonomy from the bayes table is assigned as we specified for it to be the tie breaker. 

The same thing can be applied if we want to prioritize the idtax table.

```{r, warning=FALSE}
cbind(rbind(as.data.frame(consensus.res[81, c(1:2, 4:11)]),
      cbind(consensus.res[81, 1:2], consensus.res[82, 4:11]),
      cbind(consensus.res[81, 1:2], consensus.res[83, 4:11]),
      consensus_tax_mostCom(as.data.frame(consensus.res[81, c(1:2, 4:11)]), 
                       cbind(consensus.res[81, 1:2], consensus.res[82, 4:11]), 
                       cbind(consensus.res[81, 1:2], consensus.res[83, 4:11]),
                       tablenames=c('bayes', 'idtax', 'lca'), 
                       tiebreakz=list(c('idtax', NA)), count.na=TRUE)),consensus.res[81:84, 12])
```

We can prioritize specific pairs of taxonomy and its correspnoding taxonomy table by replacing the NA with taxonomies we are interested in. If we want to prioritize the name "NA", we would put in "na".

For example, we'll prioritize the bayes table whenever we encounter the taxonomy Cryptosporidiidae. One thing to note that the tiebreaking is done by exact name matching which means that it is also case sensitive.

```{r, warning=FALSE}
cbind(rbind(as.data.frame(consensus.res[81, c(1:2, 4:11)]),
      cbind(consensus.res[81, 1:2], consensus.res[82, 4:11]),
      cbind(consensus.res[81, 1:2], consensus.res[83, 4:11]),
      consensus_tax_mostCom(as.data.frame(consensus.res[81, c(1:2, 4:11)]), 
                       cbind(consensus.res[81, 1:2], consensus.res[82, 4:11]), 
                       cbind(consensus.res[81, 1:2], consensus.res[83, 4:11]),
                       tablenames=c('bayes', 'idtax', 'lca'), 
                       tiebreakz=list(c('bayes', 'Cryptosporidiidae')), count.na=TRUE)),consensus.res[81:84, 12])
```

We can see here that in the genus, family, and order ranks, we encourted three way ties. However, in the family rank, it contained the taxonomy Cryptosporidiidae in the ties, so we prioritized it to be the tiebreaker. There was no tiebreaking specified for the others, so by default, it resorted to NA. 

### Third run

Here, we'll introduce more complex tiebreaking inputs that the algorithm can execute. Remember that the priorities are expressed based on the order of the pairs of table name and taxonomy in the list. If a certain taxonomy table is prioritized, its pair would be (table name, NA). If the specific taxonomy name "NA" wants to be prioritized, then its pair would be (table name, "na").

For example, let's build on what we had earlier and combine our tie breaking priorities. We will always prioritize the specific pair of ('bayes', 'Cryptosporidiidae') and then the idtax table itself.

```{r, warning=FALSE}
cbind(rbind(as.data.frame(consensus.res[81, c(1:2, 4:11)]),
      cbind(consensus.res[81, 1:2], consensus.res[82, 4:11]),
      cbind(consensus.res[81, 1:2], consensus.res[83, 4:11]),
      consensus_tax_mostCom(as.data.frame(consensus.res[81, c(1:2, 4:11)]), 
                       cbind(consensus.res[81, 1:2], consensus.res[82, 4:11]), 
                       cbind(consensus.res[81, 1:2], consensus.res[83, 4:11]),
                       tablenames=c('bayes', 'idtax', 'lca'), 
                       tiebreakz=list(c('bayes', 'Cryptosporidiidae'), c('idtax', NA)),
                       count.na=TRUE)),consensus.res[81:84, 12])
```

From eariler runs, we noticed that the ranks: order, family, and genus encountered three way ties. The genus and order ranks got resolved by prioritizing the idtax table. On the other hand, in the family rank, since one of the ties was Cryptosporidiidae from the bayes table, it got prioritized over the idtax table due to the order of pairs we specified in the list. 

Overall, a good tiebreak input would contain specific pairs of table and taxonomy and always contain a table to prioritize over in the end to avoid NA assignment. 

### Fourth Run

Finally, weights can be introduced into the algorithm to mitigate any ties. This is also another way to introduce bias toward taxonomy table when resolving a consensus taxonomy. Here, we'll introduce a greater weight on the bayes taxonomy table where it will be inputted twice in the pool of taxonomies to resolve a consensus to.

```{r, warning=FALSE}
cbind(rbind(as.data.frame(consensus.res[81, c(1:2, 4:11)]),
      cbind(consensus.res[81, 1:2], consensus.res[82, 4:11]),
      cbind(consensus.res[81, 1:2], consensus.res[83, 4:11]),
      consensus_tax_mostCom(as.data.frame(consensus.res[81, c(1:2, 4:11)]), 
                       cbind(consensus.res[81, 1:2], consensus.res[82, 4:11]), 
                       cbind(consensus.res[81, 1:2], consensus.res[83, 4:11]),
                       tablenames=c('bayes', 'idtax', 'lca'), 
                       count.na=TRUE, weights=c(2,1,1))),consensus.res[81:84, 12])
```

Here, we noticed that the resolved taxonomic assignments is just the bayes table.

### Fifth Run

Lastly, we can make the algorithm not consider NA as a majority candidate. 

```{r, warning=FALSE}
cbind(rbind(as.data.frame(consensus.res[61, c(1:2, 4:11)]),
      cbind(consensus.res[61, 1:2], consensus.res[62, 4:11]),
      cbind(consensus.res[61, 1:2], consensus.res[63, 4:11]),
      consensus_tax_mostCom(as.data.frame(consensus.res[61, c(1:2, 4:11)]), 
                       cbind(consensus.res[61, 1:2], consensus.res[62, 4:11]), 
                       cbind(consensus.res[61, 1:2], consensus.res[63, 4:11]),
                       tablenames=c('bayes', 'idtax', 'lca'), 
                       count.na=FALSE)),consensus.res[61:64, 12])
```

Based on the parameters of our algorithm, the consensus taxonomic assignment may contain non-assignments in the middle of the ranks which may be undesired, so user should be cautious in what parameters they set. To reduce the occurrence of these resulting taxonomic assignments, we recommend include specific tiebreaks to avoid the algorithm setting a non-assignment whenever they encounter a tie it can't resolve. Overall, a good tiebreak input would contain specific pairs of table and taxonomy and always contain a table to prioritize over in the end to avoid NA assignment.