---
title: "taxmapper_demo"
author: "D Catlett & K Son"
date: "4/13/2020"
output: github_document
---

## Overview

Here we step through an example use of the taxmapper.R function. Given a taxonomy table, it maps it to another taxonomy by exact name-matching, regardless of rank. Requires that all taxonomy tables contain sVN and ASV's at the first two columns. Synonyms for certain taxonomies can be provided to check if exact name matching doesn't work. To resolve nonexistent taxonomies from either taxonomy tables, the function assumes that `exceptions` covers all nonexistent taxonomies in the highest rank, so it can populate the rest of the ranks with NA regardless.

## Start 'er up:

We'll clear out our environment, set our wd, and read in taxonomy tables: The taxonomy tables used here come from implementations of the RDP Bayesian classifier, the newer idtaxa algorithm, and MEGAN's LCA algorithm against both the Silva and pr2 reference databases. Our amplicon data set is an 18S-V9 tag sequencing project from the coastal ocean.

You can do this with any taxonomy tables assuming you format them properly. To follow along with this demo, grab the taxonomy tables in the "test_data" directory of this repository and follow the code below.

```{r}
rm(list = ls())

# setwd and read in your datasets:
setwd("~/Desktop/Taxonomic Sequencing/amplicon_bioinformatics/tax_pipe_Mar20/")
dd <- "~/Desktop/Taxonomic Sequencing/amplicon_bioinformatics/package_deal/"

# trait and tax mapping:
source(paste0(dd,"traitmapper_Ramond.R"))
source(paste0(dd,"taxmapper.R"))
source(paste0(dd,"pr2_tax_miner.R"))
source(paste0(dd,"analyze_traitmap_byTrait.R"))

# helper fcns:
source(paste0(dd,"LCA2df.R"))
source(paste0(dd,"idtax2df_pr2.R"))
source(paste0(dd,"bayestax2df.R"))
source(paste0(dd,"idtax2df_silva.R"))
source(paste0(dd,"find_asvs_by_name.R"))

# ensemble tax algrithms:
source(paste0(dd,"consensus_tax_LCAlike.R"))
source(paste0(dd,"consensus_tax_bestRez.R"))
source(paste0(dd,"compare_taxrez.R"))

bayes.pr2 <- readRDS("initial_tax_tabs/bayes_pr2_0boot_Mar20.rds")
idtax.pr2 <- readRDS("initial_tax_tabs/idtax_pr2_0boot_Mar20.rds")
bayes.silva <- readRDS("initial_tax_tabs/bayes_silva_0boot_Mar20.rds")
idtax.silva <- readRDS("initial_tax_tabs/idtax_silva_0boot_Mar20.rds")
lca.pr2 <- readRDS("initial_tax_tabs/LCA_pr2_rawdf_Mar20.rds")
lca.silva <- readRDS("initial_tax_tabs/LCA_silva_rawdf_Mar20.rds")
```

## Arranging and formatting our taxonomy tables:

The data we're using was pulled slightly haphazardly, so here we'll use some bootstrapping estimats to NA-out low-confidence assignments, reformat our taxonomy tables as data frames, and sort them alphabetically by ASV sequences so that the order of rows/ASVs is the same across all taxonomy tables.

```{r}
# here's the rubric for aligning ASV numbers and sequences across datasets:
library("DECIPHER")
setwd("~/Desktop/Taxonomic Sequencing/amplicon_bioinformatics/tax_pipe_Mar20/")
rubber <- readDNAStringSet("blaster/allASVs4blast.fasta")

# convert idtax and bayes to dataframes - keep boot = 0  and return conf for now...
# note that for idtaxa, order of ASVs must be retrieved from dada2 seqtab - rubric represents that order
# and adds ASV names and sequences to output dataframe:
xx <- bayestax2df(bayes.pr2, boot = 0, rubric = rubber, return.conf = TRUE)
bayes.pr2 <- xx[[1]]
bayes.pr2.conf <- xx[[2]]

xx <- bayestax2df(bayes.silva, boot = 0, rubric = rubber, return.conf = TRUE)
bayes.silva <- xx[[1]]
bayes.silva.conf <- xx[[2]]

xx <- idtax2df_pr2(idtax.pr2, boot = 0, rubric = rubber, return.conf = TRUE)
idtax.pr2 <- xx[[1]]
idtax.pr2.conf <- xx[[2]]

xx <- idtax2df_silva(idtax.silva, boot = 0, rubric = rubber, return.conf = TRUE)
idtax.silva <- xx[[1]]
idtax.silva.conf <- xx[[2]]

# re-order your 6 tables by sorting according to ASV:
ii <- base::sort(bayes.pr2$ASV, index.return = TRUE)
bayes.pr2 <- bayes.pr2[ii$ix,]
bayes.pr2.conf <- bayes.pr2.conf[ii$ix,]

ii <- base::sort(idtax.pr2$ASV, index.return = TRUE)
idtax.pr2 <- idtax.pr2[ii$ix,]
idtax.pr2.conf <- idtax.pr2.conf[ii$ix,]

ii <- base::sort(lca.pr2$ASV, index.return = TRUE)
lca.pr2 <- lca.pr2[ii$ix,]
# no confidence for lca tax assignments...

ii <- base::sort(bayes.silva$ASV, index.return = TRUE)
bayes.silva <- bayes.silva[ii$ix,]
bayes.silva.conf <- bayes.silva.conf[ii$ix,]

ii <- base::sort(idtax.silva$ASV, index.return = TRUE)
idtax.silva <- idtax.silva[ii$ix,]
idtax.silva.conf <- idtax.silva.conf[ii$ix,]

ii <- base::sort(lca.silva$ASV, index.return = TRUE)
lca.silva <- lca.silva[ii$ix,]
```

You can run this for a sanity check:

```{r, eval=FALSE}
# check that they're all in the same order
identical(bayes.pr2$ASV, bayes.silva$ASV)
identical(bayes.pr2$ASV, idtax.pr2$ASV)
identical(idtax.pr2$ASV, idtax.silva$ASV)  
identical(idtax.silva$ASV, lca.pr2$ASV)
identical(lca.pr2$ASV, lca.silva$ASV)
```

...and this to see what the data sets look like. These data sets are available in the test-data directory.

```{r, eval=FALSE}
head(bayes.pr2)
head(bayes.silva)
head(idtax.pr2)
head(idtax.silva)
head(lca.pr2)
head(lca.silva)
```

## Run taxmapper 

Here we'll do an example run and see what the inputs and outputs look like. The function iterates through each row of the `taxin` data frame to map each row to a row from `tax2map2` data frame. The function returns a list of three elements. The first element is a data frame including all unique taxonomic assignments and their corresponding mapped assignments. The second element is a character vector containing all names within `taxin` that were unable to be mapped. The last element is a data frame having the ASV seqs supplied in `taxin` with the mapped taxonomic assignments.

In this example, we'll map bayes-silva to bayes-pr2:

```{r}
source("~/Desktop/Taxonomic Sequencing/amplicon_bioinformatics/taxonomy_pipeline/tax_table_mapping/taxmapper.R")

synonym.filepath <- "~/Desktop/Taxonomic Sequencing/amplicon_bioinformatics/taxonomy_pipeline/tax_table_mapping/tax_synonyms_FINAL.csv"

# Bacteria and Archaea doesn't exist in pr2
nonexistent <- c('Bacteria', 'Archaea')

pr2 <- read.csv("~/Desktop/Taxonomic Sequencing/amplicon_bioinformatics/taxonomy_pipeline/tax_table_mapping/pr2_all_tax.csv")
pr2 <- pr2[,-1]

bayes.silva.2.pr2 <- taxmapper(taxin=bayes.silva, tax2map2=pr2, 
                               exceptions=nonexistent,
                               synonym.file=synonym.filepath)
```

`bayes.silva.2.pr2` contains a list of the three outputs discussed earlier. Let's look through each element, starting with the unique taxonomic assignments from bayes-silva to bayes-pr2's corresponding mapped assignments.

```{r}
head(bayes.silva.2.pr2[[1]], 20)
```

Here we see any instances of Archaea or Bacteria from bayes-silva gets mapped to Bacteria on the bayes-pr2 side.

Let's look at the character vector of taxonomies from bayes-silva that wasn't able to be mapped.

```{r}
head(bayes.silva.2.pr2[[2]], 10)
```

Let's look at the data frame of ASV's from bayes-silva mapped to their corresponding taxonomies from bayes-pr2.

```{r}
head(bayes.silva.2.pr2[[3]], 20)
```